### PR TITLE
(OPTIMIZE=2, Unstable) AutoScheduler (GEMM > 300GFLOP/s on CPU, > 4000 GFLOPs on GPU)

### PR DESCRIPTION
**Pull request by hikettei**
### What's new?

- [ ] Break into multiple PR ...
- [ ] Or reimplement everything in codegen ...
- [ ] DO A CODEGEN REFACTOR FIRST!!
- [x] **OPTIMIZE=1** will generate a fast kernel without profiling the actual execution. (i.e.: effective for dynamic shape)
- [x] **Sketch** Caten will generate a sketch, an optimizing kernel template which is generated by solving ILP
- [x]  **Memory Optimization** Caten can optimize the memory layout for any architectures specified by users.
  - [x] For CPU, aligned/packed memory fitting in L1 Cache
  - [x] For Metal, simdgroup
  - [ ] For CUDA, shared memory
- [ ] **Vectorize** Caten can fully utilize a microkernel specified by a user
  - [ ] SIMD for 26 ops is all you need
  - [ ] TensorCore
- [ ] **BEAM Search** Caten will optimize the parameter by profiling the execution time (during computation!!!)
  - [ ] All you have to optimize during the compilation is the following:
    - [ ] BLOCKSIZE
  - [ ] During the computation, caten will optimize:
    - [ ] LocalSize in GPU
- [ ] **OPTIMIZE=2** will sample multiple sketch during the compilation, selecting the best one.
- [ ] 🎉 Release v0.3

image

```
(Sketch Example for CPU gemm)
#<SKETCH 
  // @DIRECTIVE(TYPE=PARALLEL,AMOUNT=(0),VISIBLE=T)
  for (int c0 = 0; c0 < a; c0 += 8)
    for (int c1 = 0; c1 < a; c1 += 8) {
      // @DIRECTIVE(TYPE=PACKED_OUTER,AMOUNT=(4 4),VISIBLE=T)
      for (int c2 = 0; c2 <= min(7, a - c0 - 1); c2 += 4)
        for (int c3 = 0; c3 <= min(7, a - c1 - 1); c3 += 4) {
          // @DIRECTIVE(TYPE=PACKED_INNER,AMOUNT=(4 4),VISIBLE=NIL)
          for (int c4 = 0; c4 <= min(3, a - c0 - c2 - 1); c4 += 1)
            for (int c5 = 0; c5 <= min(3, a - c1 - c3 - 1); c5 += 1) {
              // @DIRECTIVE(TYPE=PACKED_OUTER,AMOUNT=(4),VISIBLE=T)
              {
                // @DIRECTIVE(TYPE=PACKED_INNER,AMOUNT=(4),VISIBLE=NIL)
                NID1876621(c0 + c2 + c4, c1 + c3 + c5);
                for (int c6 = 0; c6 < a; c6 += 4) {
                  // @DIRECTIVE(TYPE=PACKED_INNER,AMOUNT=(4),VISIBLE=NIL)
                  for (int c7 = 0; c7 <= min(3, a - c6 - 1); c7 += 1)
                    NID1876636(c0 + c2 + c4, c1 + c3 + c5, c6 + c7);
                }
                // @DIRECTIVE(TYPE=PACKED_INNER,AMOUNT=(4),VISIBLE=NIL)
                NID1876641(c0 + c2 + c4, c1 + c3 + c5);
              }
            }
        }
    }
  :search-params NIL
  :opt-history ((0 . #<PACKING :id NIL :amount 4>)
                (0 . #<PACKING :id NIL :amount 4>)
                (0 . #<TILEBAND :id NIL :amount (8 8)>)
                (0 . #<PARALLEL :id NIL :amount NIL>)
                (0 . #<RESCHEDULE :id NIL :amount NIL>))
 {7031D1E363}>
```

- [ ] define-auto-scheduler
  - [ ] Add: SupportPrefetch (and the size)
  - [ ] Add: SupportVectorize
- **CPU 300 GFlops Workload**
  - [x] Move PREFETCH to the outermost band
  - [x]  Collapse(2) is doable from band-depth
  - [ ] Rewriting rule: LOAD, STORE ==> Caching mutation
  - [ ] ReduceOpsのFirst OpsかどうかでCacheするかしないかが決まる
  - [ ] Ifを挿入してLOADかRestoreかを選ぶ
  - [ ] Activation Fusion
  - [x] ast-rewrite-zero-indexing (AstForを全て0 Startにする(make-astforするときに))
  - [ ] If --> AllowGIDRewriter option and move
  - [ ] Full Symbolic is a little slower
    - [ ] **Tiled Dimの共通項削除が必要** (max(...))
- **GPU 4000 GFLOPs Workload**
  - [ ] simdgroup
  - [ ] simdgroup_load
  - [ ] TensorCore WMMA
  - [ ] singgroup_store
  - [ ] 4000GFLOP/s
- **Prefetch Directive Workload**
  - [ ] A compiler will insert :LOAD and :STORE
  - [ ] A default load and store (e.g.: activation) is later inserted considering the cache
- [ ] TILE_OUTER+TILE_INNER duplication is purged and removed.
- [ ] implement everything in the figure and pass all tests with OPTIMIZE=1 (default behaviour)
  - [ ] Able to optimize full symbolic kernels
- [x] Our first priority is to provide decent schedule commands, and achieve OpenBLAS/cuDNN-level performance via with-manual-scheduling
  - [ ] The key idea is the microkernel transformation
- [x] Always use with-manual-scheduling instead of doing BFS in ISL Schedule Node.
- [x] Add: blueprint verifier
- [x] Solving ILP is always safe
- [x] (Stage1) Complete Polyhedral Transformation (in transform.lisp)
  - [x] Is this set of optimizations enough to generate a theoretically optimal kernel?  
  - [x] UNROLL
    - [x] Test: val_x_y indexing is 100% working?
  - [x] PACKING
    - [x] Test: val_x_y indexing is 100% working?
   - [x] TILE
   - [x] ~~TILE_ND~~
     - [x] Sink the bands as possible
     - [x] Also use ISL Scheduler
     - [x] ensure the new schedule fits in a single kernel
   - [x] PARALLEL
     - [x] Legality: Inner softmax bands are not coincident! 
   - [x] GLOBAL
     - [x] Implemented as a tile
   - [x] LOCAL
   - [ ] APPLY_GLOBAL_COLEASING
     - [x] How to ensure K-warp is applied? by profiling?
   - [x] APPLY_CACHE_BLOCKING
- [ ] (Stage2) ISL AST -> Lisp AST Lowerer
  - [x] Multiple Bands: Splitして何通りか候補を生成 -> Sort
  - [x] Refactor:
    - [x] parse-isl-markは直下のASTForのMarkにDirectiveをMappingさせるだけにする(Directive+BANDは100%)
    - [x] ASTをParseした後に，Directiveをもとに変形する処理を実装する (処理を2段階に分ける)
    - [x] amount=list
  - [x] UNROLL
    - [x] TODO: Padding the user-args with the maximum length of unroll space
  - [x] PACKING
    - [x] Check: LayerNorm 
  - [x] TILE
  - [x] TILE_ND
  - [x] PARALLEL
  - [x] GLOBAL
    - [x] Lowered as: const idx_x = blockIdx.x * ls
  - [x] LOCAL
    - [x] Lowered as: const idx_y = threadIdx.y
  - [ ] APPLY_GLOBAL_COALESCING
    - [ ] Implemented by rewriting schedule_node_band
  - [x] APPLY_CACHE_BLOCKING
    - [x] Implement: BARRIER
- [ ] (Stage3) User-defined microkernel transformation by PACKING
    - [ ] We can generate the optimal gemm kernel because caten can replace the user-defined micro kernel.
    - [ ] Microkernel includes: FLOAT4, SIMD(__mm256), gemmAxB
    - [ ] (Optional) gemmAxB is generated by with-manual-scheduling in simd.lisp
    - [ ] Complete TensorCore
      - [ ] Insert :PACK :UNPACK renderer
      - [ ] Implement all mutation pattern for 26 math ops for AVX2, ArmNeon.
      - [ ] Both should be tested by CI (CLANG[AVX2] CLANG[ARMNEON])
- [ ] 一旦最適化を完成させる:
  - [ ] TensorCore
  - [ ] Float4, vectorize
  - [ ] Memory Coalescing (K-Warp)
  - [ ] Parameter Tuning (TILE_SIZE, LOCAL)
  - [x] Shared Memory Blocking
  - [ ] ASTRewrite without making a copy
  - [ ] Optimize the whoe things on ./auto-scheduler (set optimize 3)
- [ ] (Stage4) Sketch Parameter OPTIMIZATION with OPTIMIZE=2
  - [ ] Search spaces are: LocalSize, TileSize, etc
- [ ] gemm 120 GFLOPS on CPU, as well as gpu (1500gflops on m3 pro?)
- [ ] Complete ./docs/byoc.lisp ./docs/gemm.lisp
- [ ] Complete env.lisp
  - [ ] Detect the arch of SIMD used by CLANG backend, by adding cffi-grovel file.
- [ ] (Stage4) OPTIMIZE=1 can be implemented as:
```lisp
(with-manual-scheduling (kernel BYOC_AUTO_SCHEDULER)
  (dotimes (I (n-bands))
     (opt (make-instance 'Packing :amount 4) I))
  ;; ...
)
```
- [ ] (Stage5) OPTIMIZE=2 can be implemented as BEAM Search (coming in the next PR)

## todo 

- [x] :GLOBAL is a tile
- [ ] one sized loop cannot be parallelized or insert visible=nil when doing scop.
- [ ] Pass all tests w/ METAL (why 4d matmul returns segv??)
- [x] WMMA/TC/Vectorize
- [x] Still looking for Smth between Polyhedral + BEAM Search ...
  - [x] https://arxiv.org/pdf/2006.06762
  - [x] ppcg https://dl.acm.org/doi/pdf/10.1145/2400682.2400713
  - [x] https://github.com/facebookincubator/AITemplate/wiki/How-to-write-a-fast-Softmax-CUDA-kernel%3F
  - [x] Diesel: DSL for linear algebra and neural net computations on GPUs
  - [x] maybe i should start by generating hand scheduled gemm/softmax/conv kernel 
  - [x] Add: New JITAble OP; :CUSTOM

## Workload

- [ ] Auto parallelization for GPU kernel (passing argmax failing case, softmax)
- [ ] hand optimized gemm kernel achieving 100 GFLOPS on M3 Pro+gcc
- [ ] hand optimized gemm kernel with Metal
- [x] Implement `TensorCore` which optimizes MULADD computation:
  - [ ] CPU Gemm: https://salykova.github.io/matmul-cpu
    - [x] Feed this microkernel to the clang backend https://github.com/salykova/matmul.c/blob/main/src/kernel.c#L3
    - [ ] Add caten/byoc/env.lisp
      - [ ] Selects the best backend for each computers
      - [ ] (print-env) to show the env (e.g.: AVX2/AVX512 is available? etc)
      - [ ] ./docs/bring-your-own-codegen.lisp
      - [ ] ./docs/how-to-optimize-gemm.lisp

This PR firstly introduces manually scheduling commands until achieving 90% performance of OpenBLAS on CLANG:

```lisp
(deftest hand-optimized-cpu-gemm-test
  (let ((raw (get-gemm-schedule)))
    (with-manual-scheduler (raw Mock-CPU-AutoScheduler)
      ;; Scheduling Priority:
      ;; Interchange(Memory Layout) -> Packing -> Tile -> Interchange (2D Tile) -> Parallelize
      ;; (Caten will automatically find TensorCore if the bands are packed)
      (opt (make-instance 'Packing :amount 4) 0)
      (opt (make-instance 'Packing :amount 4) 1)
      (opt (make-instance 'Packing :amount 4) 2)
      ;; 2D Tiling (16, 16)
      ;; (opt (make-instance 'TileBand :amount 16) 0)
      ;; (opt (make-instance 'TileBand :amount 16) 1)
      ;; (opt (make-instance 'TileBand :amount 16) 2)
      ;; (opt (make-instance 'Interchange :amount 1) 1)
      ;; Parallelize at innermost dimension
      ;; (opt (make-instance 'Parallel) 5)
      )
    (print-schedule raw)
    (print-bp raw)))
```

Once this pr is merged auto scheduler is just finding a sequence of opt.

- **TensorCore workload**
  - [x] Implement Packing
    - [x] packing.lisp: AST Level rewriter for the vectorize. If failed, replaced with unroll.
    - [x] AST Level rewriting for packing (if failed, replace with unroll)
  - [x] define-auto-scheduler will receive TensorCore
  - [x] Replace TensorCore.shape packing w/ Tensorcore
  - [ ] Add: failing to vectorized case (add global-space)
  - [ ] Add: TensorCore Renderer
  - [ ] Make unroll/packing stable for all kernel
  - [ ] CLANG: ArmNeon, AVX2, AVX512, gemm8x8 (but I don't have intel arch computer now?+how to test this in ci?)
    - [ ] CI: CLANG[AVX2], CLANG[ArmNeon], CLANG[AVX512], ClangRuntime will automatically detect the archtecture
    - [ ] Add: solve-pack-unpaxk
    - [ ] expr meta: dont forget to create a copy of node first 
- [ ] OPTIMIZE=1 Strategy(CPU):
  - [ ] Applying PACKING(4) for all dims (applying gemm8x8 microkernel)
  - [ ] Applying Tile(Constant) for all dims
  - [ ] Select coincident dims and apply parallel
  - [ ] Use Pointer incf instead of a[...]
  - [ ] Support SLEEF Vectorize
  - [ ] LISP TENSORCORE? (implement hand optimized gemm4x4?...)
  - [x] Gemm 90 GFlops on gcc
  - [x] Gemm 120 GFlops on gcc
  - [x] Gemm 150 GFlops on gcc
  - [ ] Gemm 3900 GFLOPS on Metal

### CODE REVIEW

**Comment by hikettei** [comment]
Template for Clang TensorCore gemm (40GFlops w/o writing simd intrinsic directly, GCC O3)

```c
static inline void _avx2_gemm8x8_float_tc(float *a, float *b, float *c, int mr, int nr, int kc, int m) {
  #pragma omp unroll(4)
  for (int i = 0; i < 4; i++) {
    #pragma omp unroll(4)
    for (int j = 0; j < 4; j++) {
      float sum = 0.0f;
      #pragma omp unroll(4)
      for (int k = 0; k < 4; k++) {
        sum += a[i * kc + k] * b[k * nr + j];
      }
      c[i * m + j] += sum;
    }
  }
}

void gemm(int M, int N, int K,
          const float * __restrict A,
          const float * __restrict B,
          float * __restrict C)
{
  #pragma omp parallel for collapse(2)
  for (int _gid0=0;_gid0<512;_gid0+=4) {
    for (int _gid2=0;_gid2<512;_gid2+=4) {
      //float val_2_0 = 0.0;
      //float val_2_1 = 0.0;
      //float val_2_2 = 0.0;
      //float val_2_3 = 0.0;
      for (int _gid4=0;(_gid4<512);_gid4+=4) {
        _avx2_gemm8x8_float_tc((float*)&A[((512*_gid0)+_gid4)], (float*)&B[(_gid2+(512*_gid4))], (float*)&C[((512*_gid0)+_gid2)], 512, 512, 512, 512);
      }
      //C[((512*_gid0)+_gid2)] = val_2_0;
      //C[((512*_gid0)+(_gid2+1))] = val_2_1;
      //C[((512*_gid0)+(_gid2+2))] = val_2_2;
      //C[((512*_gid0)+(_gid2+3))] = val_2_3;
    }
  }
}
```

**Comment by hikettei** [comment]
Currently 120 GFlops on cpu, to make it faster than OpenBLAS, we also have to add:
- [ ] Microkernel rewriting
- [ ] Parameter Fine Tuning (TILE_SIZE, PACK_SIZE)
- [ ] SIMD Rewriting

image

**Comment by hikettei** [comment]
実験色々回して共通する方針の気持ちがわかってきたような？
- 1. 128x128くらいで一番外側のBandをTileする
- 2. Prefetch
  - GPUならShared MemoryへCopy+Barrier
  - CPUならstatic float ... __attribute__((aligned(64)));した固定の領域にデータを移動 + この直下にPARALLELを挿入
- 3. Compute
  - Prefetchした領域内で計算を回す (SIMD化できるようにAlignされてるはず)

**Comment by hikettei** [comment]
90% perf of OpenBLAS is ok (350GFLOPS); lets working on:
- Pass all test w/ CPU (add activation fuse with caching rewriting rule)
- Metal

**Comment by hikettei** [comment]
### Metal Gemm Baseline

- [ ] Metal 4000 GFLOP/s
  - [ ] Rewrite threadgroup -> simdgroup
  - [ ] _gid5=0; loop -> simdgroup_load
  - [ ] the innermost loop -> Rewrite as TensorCore

```c
kernel void fused_sumnode_matmul252065(device float *val_9, device const float *val_6, device const float *val_4, uint3 gid [[threadgroup_position_in_grid]], uint3 lid [[thread_position_in_threadgroup]]) {
  int64_t _gid0 = (gid.x*8); /* GRID_0<#<EXPR 64>, #<EXPR 8>> */
  int64_t _gid1 = (gid.y*8); /* GRID_1<#<EXPR 64>, #<EXPR 8>> */
  int64_t _gid2 = lid.x;
  int64_t _gid3 = lid.y;
  threadgroup float _val_4_s[4096];
  threadgroup float _val_6_s[4096];
  for(int _gid4=0;(_gid4<512);_gid4+=64) {
   for(int _gid5=0;(_gid5<64);_gid5+=1) {
     (*(_val_4_s+((64*_gid2)+_gid5))) = (*(val_4+((512*(_gid0+_gid2))+(_gid4+_gid5))));
    }
    for(int _gid5=0;(_gid5<64);_gid5+=1) {
      (*(_val_6_s+((64*_gid3)+_gid5))) = (*(val_6+((_gid1+_gid3)+(512*(_gid4+_gid5)))));
    }
    threadgroup_barrier(mem_flags::mem_threadgroup);
    float val_2_0_0_0 = 0.0;
    for(int _gid5=0;(_gid5<64);_gid5+=1) {
      val_2_0_0_0 = (val_2_0_0_0+((*(_val_4_s+((64*_gid2)+_gid5)))*(*(_val_6_s+((64*_gid3)+_gid5)))));
     }
     (*(val_9+((512*(_gid0+_gid2))+(_gid1+_gid3)))) = val_2_0_0_0;
    }
  }
}
```
